File: leetcode/tree/binary_search_tree_iterator.py
from common import TreeNode

class BSTIterator:

    def __init__(self, root: TreeNode):
        self.arr = []
        self.helper(root)        

    def helper(self, root):
        if not root:
            return
        self.helper(root.left)
        self.arr.append(root.val)
        self.helper(root.right)
        
    def next(self) -> int:
        """
        @return the next smallest number
        """
        return self.arr.pop(0)        
        

    def hasNext(self) -> bool:
        """
        @return whether we have a next smallest number
        """
        return len(self.arr) != 0

if __name__ == "__main__":
    root = TreeNode(7)
    root.left = TreeNode(3)
    root.right = TreeNode(15)
    root.right.left = TreeNode(9)
    root.right.right = TreeNode(20)
    bst = BSTIterator(root)
    for _ in range(5):
        print(bst.next())
        print(bst.hasNext())

File: leetcode/tree/binary_search_tree_iterator.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type BSTIterator struct {
	Arr []int
}

func Constructor(root *TreeNode) BSTIterator {
	bst := BSTIterator{}
	bst.helper(root)
	return bst
}

func (this *BSTIterator) helper(root *TreeNode) {
	if root == nil {
		return
	}
	this.helper(root.Left)
	this.Arr = append(this.Arr, root.Val)
	this.helper(root.Right)
}

/** @return the next smallest number */
func (this *BSTIterator) Next() int {
	tmp := this.Arr[0]
	this.Arr = this.Arr[1:]
	return tmp
}

/** @return whether we have a next smallest number */
func (this *BSTIterator) HasNext() bool {
	return len(this.Arr) != 0
}

func main() {
	root := &TreeNode{Val: 7, Left: nil, Right: nil}
	root.Left = &TreeNode{Val: 3, Left: nil, Right: nil}
	root.Right = &TreeNode{Val: 15, Left: nil, Right: nil}
	root.Right.Left = &TreeNode{Val: 9, Left: nil, Right: nil}
	root.Right.Right = &TreeNode{Val: 20, Left: nil, Right: nil}

	bst := Constructor(root)
	for i := 0; i < 5; i++ {
		fmt.Println(bst.Next())
		fmt.Println(bst.HasNext())
	}
}
